query/src/sql/planner/binder/sort.rs

```diff
@@ -13,42 +13,58 @@
 // limitations under the License.
 
 use common_ast::ast::OrderByExpr;
+use common_ast::parser::error::DisplayError;
+use common_exception::ErrorCode;
 use common_exception::Result;
 
 use crate::sql::binder::scalar::ScalarBinder;
 use crate::sql::binder::Binder;
 use crate::sql::optimizer::SExpr;
+use crate::sql::plans::Scalar;
 use crate::sql::plans::SortItem;
 use crate::sql::plans::SortPlan;
 use crate::sql::BindContext;
 
 impl<'a> Binder {
     pub(super) async fn bind_order_by(
         &mut self,
+        child: SExpr,
         order_by: &[OrderByExpr<'a>],
-        bind_context: &mut BindContext,
-    ) -> Result<()> {
-        let scalar_binder = ScalarBinder::new(bind_context, self.ctx.clone());
-        let mut order_by_exprs = vec![];
-        for expr in order_by {
-            order_by_exprs.push(scalar_binder.bind_expr(&expr.expr).await?);
-        }
-
-        let mut order_by_items = Vec::with_capacity(order_by_exprs.len());
-        for (idx, order_by_expr) in order_by_exprs.iter().enumerate() {
+        input_context: &BindContext,
+        output_context: &BindContext,
+    ) -> Result<SExpr> {
+        let select_scalar_binder = ScalarBinder::new(output_context, self.ctx.clone());
+        let from_scalar_binder = ScalarBinder::new(input_context, self.ctx.clone());
+        let mut order_by_items = vec![];
+        for order in order_by {
+            // First we try to resolve sort item with `SELECT` context
+            let mut res = select_scalar_binder.bind_expr(&order.expr).await;
+            // If failed, we will try to resolve sort item with `FROM` context
+            if let Err(e) = res.clone() {
+                // If still failed, the previous error will be raised
+                res = from_scalar_binder.bind_expr(&order.expr).await.or(Err(e));
+            }
+            let (scalar, _) = res?;
+            if !matches!(scalar, Scalar::BoundColumnRef(_)) {
```

Comment:
Similar statements such as `select a from t order by a + b` don't seem to be allowed here?

Reply:
It's unclear. We will only support order by columns can be displayed in result set.